python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,520 @@
+
+    def can_transform(self, target: Type) -> bool:
+        if isinstance(self._type, FixedType):
+            return target == BinaryType or isinstance(target, FixedType)
+        elif isinstance(self._type, DecimalType):
+            return isinstance(target, DecimalType)
+
+        return Bucket._FUNCTIONS_MAP[self._type][1](target)
+
+    def get_result_type(self, source: Type):
+        return IntegerType
+
+    def to_human_string(self, value):
+        return str(value)
+
+
+class Time(Transform):
+    """
+    Time class is for both Date transforms and Timestamp transforms.
+    """
+
+    _TIME_ORDER = datetime(year=3, month=2, day=1, hour=0)
+    _VALID_TIME_GRANULARITY = {
+        DateType: {"year", "month", "day"},
+        TimestampType: {"year", "month", "day", "hour"},
+        TimestamptzType: {"year", "month", "day", "hour"},
+    }
+    _DIFF_MAP = {
+        "year": lambda t1, t2: (t1.year - t2.year)
+        - (
+            1
+            if t1.month < t2.month or (t1.month == t2.month and t1.day < t2.day)
+            else 0
+        ),
+        "month": lambda t1, t2: (t1.year - t2.year) * 12
+        + (t1.month - t2.month)
+        - (1 if t1.day < t2.day else 0),
+        "day": lambda t1, t2: (t1 - t2).days,
+        "hour": lambda t1, t2: int((t1 - t2).total_seconds() / 3600),
+    }
+
+    def __init__(self, transform_type: Type, name: str):
+        if name not in Time._VALID_TIME_GRANULARITY.get(transform_type, {}):
+            raise ValueError(f"Cannot transform type: {transform_type} by {name}")
+
+        super().__init__(
+            name, f"transforms.{name}(transform_type={repr(transform_type)})"
+        )
+        self._type = transform_type
+        self._name = name
+
+    def apply(self, value: int) -> int:
+        if self._name == "day" and self._type == DateType:
+            return value
+
+        if self._type == DateType:
+            value_time = datetime.utcfromtimestamp(value * 86400)
+        else:
+            value_time = datetime.utcfromtimestamp(value / 1000000)
+
+        return Time._DIFF_MAP[self._name](value_time, Transform._EPOCH)
+
+    def can_transform(self, target: Type) -> bool:
+        if self._type == DateType:
+            return target == DateType
+        else:  # self._type is either TimestampType or TimestamptzType
+            return target == TimestampType or target == TimestamptzType
+
+    def get_result_type(self, source_type: Type) -> Type:
+        if self._name == "day":
+            return DateType
+
+        return IntegerType
+
+    def preserve_order(self) -> bool:
+        return True
+
+    def satisfy_order(self, other: Transform) -> bool:
+        if self == other:
+            return True
+
+        if isinstance(other, Time):
+            return getattr(Time._TIME_ORDER, self._name) <= getattr(
+                Time._TIME_ORDER, other._name
+            )
+
+        return False
+
+    def to_human_string(self, value: int) -> str:
```

Comment:
I think most of the methods in this class should have the implementation determined in `__init__`. This is another example where you don't need to check `_name` or `_type` in the implementation.

Reply:
👌